### ISSUE TITLE
No support for lookbehind

### ISSUE THREAD

**Issue by naomik** [atom/find-and-replace#571]
Implementing JavaScript's native RegExp engine for a find/replace tool in a text editor seems like a dramatic oversight.

As find-and-replace leans on JS's RegExp, of course the following find expressions is invalid

``` js
(?<=hello)world
(?<!hello)world
```

Errors

``` plain
Invalid regular expression: /(?<=hello)world/: Invalid group
Invalid regular expression: /(?<!hello)world/: Invalid group
```

This module is unusable without such features

**Comment by Epigene** [comment]
I agree, lookaround is crucial!

Using JS's RegExp is also weird considering Oniguruma is already used within Atom to parse grammar files as mentioned in [this exchange](https://discuss.atom.io/t/regular-expression-lookaround-support-in-find/12400).

**Comment by abe33** [comment]
I as already mentioned on Discuss, using oniguruma for grammar made sense as it allows to reuse TextMate and SublimeText grammars out of the box, but AFAIK that's the only context where oniguruma is used. 
OTOH I agree it's a shame that lookbehind is not part of JS regexp (while loohahead is).

**Comment by JesseLeite** [comment]
I don't know much about the subject, apart from lookbehinds not being part of JS regex.  However, as a PHP dev who regular-ily (pun intended) regexes via PHP regex  engine with lookbehinds, it sucks not being able to lookbehind in Atom's find and replace.  Curious about this Oniguruma?

**Comment by alexchandel** [comment]
Oniguruma has this feature and many, many more. There is absolutely no reason to use JS's garbage regex engine over Oniguruma.

**Comment by JesseLeite** [comment]
@alexchandel I'm unfamiliar with Oniguruma. I'm all for more features, but is the syntax standard and testable on a site like [regex101.com](http://regex101.com)?

**Comment by 50Wliu** [comment]
@JesseLeite It's what Atom uses for languages, and I've never had a problem testing any of those regexes on regex101.com.

**Comment by bb-jakespirek** [comment]
Bummed that lookbehind doesn't work.

**Comment by alexchandel** [comment]
@bb-jakespirek It would if #698 happened

**Comment by alexchandel** [comment]
Since #698 is closed as a duplicate, is this issue now for using Oniguruma for regexes?

**Comment by naomik** [comment]
relevant: https://github.com/atom/find-and-replace/issues/570

Oniguruma would fix this too.

**Comment by bric3** [comment]
@JesseLeite https://regex101.com/r/dO0yG2/1

**Comment by lzkelley** [comment]
Would this also allow better handling of indentation errors (e.g. https://github.com/atom/language-python/issues/22)?  Which really shouldn't need special packages to solve (e.g. https://atom.io/packages/python-indent)...

**Comment by 50Wliu** [comment]
@lzkelley No, as find-and-replace has nothing to do with language grammars.  Those are already using Oniguruma.

**Comment by MattSturgeon** [comment]
> Those are already using Oniguruma.

Wait.. Atom already uses Oniguruma, but not for Find? Surely it should be a simple fix then? ;)

**Comment by ErikCorryGoogle** [comment]
Latest V8 releases have lookbehind available.  It's behind the experimental JavaScript flag.

**Comment by alexchandel** [comment]
V8 still doesn't have have async callbacks, and catastrophic backtracking and misspelt regexes are the primary source of hangs in find-and-replace, for example #557 and #856. Why not simply switch to Oniguruma and be done with it? There have been hundreds of commits since this was opened, clearly someone has the time to fix this.

**Comment by lee-dohm** [comment]
@alexchandel your comment was deleted as a violation of the [Atom Code of Conduct](https://github.com/atom/atom/blob/master/CODE_OF_CONDUCT.md) as it is insulting or derogatory. You may consider this an official warning.

**Comment by MattSturgeon** [comment]
For those interested, @alexchandel was saying that while V8 has added experimental lookbehind support, it is still inferior to many alternative regex libraries in his opinion:

> V8 still doesn't have have async callbacks, and catastrophic backtracking and misspelt regexes are the primary source of hangs in find-and-replace, for example #557 and #856.

He then went on to suggest using an alternative like [Oniguruma](https://github.com/kkos/oniguruma) (and rather rudely implied that the atom team should be prioritising this issue higher — hence the deleted comment).

**Comment by MartinBonner** [comment]
I was trying to find the documentation of the Atom regex syntax (given that they are all different), and was pleased to find that it used Oniguruma (with the lookbehind I wanted), and then was gutted to discover this bug.

It does seem rather extraordinary that Atom uses Oniguruma, but not for actual search-and-replace!

**Comment by Stanzilla** [comment]
VS Code recently [switched](https://github.com/Microsoft/vscode/commit/62ec995019e86074c5aa24003d2f960bd9282d90) to [ripgrep](https://github.com/BurntSushi/ripgrep) for search and it might be worth investigating doing the same in Atom. It is really fast. ⚡️ 

* https://github.com/roblourens/vscode-ripgrep
* https://github.com/Microsoft/vscode/issues/19983

**Comment by MartinBonner** [comment]
ripgrep is great for a "find in files" solution, _but_ it achieves that speed (at least in part) by not supporting certain functions (the other part is achieved by not looking in .gitignore'd files - which is also only relevant to "find in files".  In particular, I think it doesn't support look-around (which is what I was after).  I don't think raw speed is nearly so important for a complex search-and-replace (which is what I wanted this feature for).

**Comment by ai-danno** [comment]
As a regex nerd looking for the ultimate editor, I was drawn to this conversation.

**Question**- does Atom now support lookbehinds? Thanks in advance!

**Comment by MartinBonner** [comment]
@ai-danno : No it doesn't.  The lack of lookbehind is what drew me to this thread in the first place.  

My comment was because switching to ripgrep would give us speed, but not lookaround (and I think that would be a poor tradeoff).

**Comment by ai-danno** [comment]
@MartinBonner 👍  Thanks for the fast reply. It seems like it might bubble to the surface here in not too long, so I'll keep checking back.

**Comment by lope** [comment]
I agree 100% JS's RegExp implementation is a POS. No lookbehind is such a pain!

**Comment by lope** [comment]
So what will it take for Find/Replace to use Oniguruma instead of native JS. There's no point in waiting for new V8's lookbehind features. Oniguruma supports stuff like reset `\K` and it's working today, IN ATOM. It just needs to be USED for find/replace.

So who do we have to hassle or bribe to get this?

**Comment by ai-danno** [comment]
AS an end-user, lookbehind is the holy-grail of nerdy text-editing. Kate is free and has decent regex, but only lookahead. EditPadPro has great regex... for $50!!!

And I feel like I'm not the only one watching the capabilities of the various editors. To have this functionality in Atom, IMHO, would be a best-kept-secret advertisement for Atom. As the user, when you commit to a text-editor you do it fully, for the long-haul- you don't want to commit to a mostly-functional editor. Missing features like this become important in that decision-making process for the user.

IMHO, Atom is almost there. 
Also, it's encouraging to see discussion like this, makes me want to use it anyway.. keep up the good work!! 
:)

If my comments overstepped the bounds of this conversation or forum, I apologize.

**Comment by 50Wliu** [comment]
@lope and others: Please feel free to submit a PR!  The Atom team is primarily focusing on startup time and performance improvements at the moment, so we don't have the manpower to implement this at the moment, but we would definitely be open to a well-thought-out PR that switches to an alternative regex engine.

**Comment by lee-dohm** [comment]
As this issue stands it isn't actionable. Here's why:

* There is no definition of "better"
* There is disagreement on which features are "necessary"
* There is no list of which features people require
* There is no definition of performance expectations for both speed and memory (because I'm certain that will be the first complaint even if we select an engine that has all the features everyone wants)

Any work the maintainer team would do along these lines would ultimately be wasted effort. Even if someone submitted a PR using a different regular expression engine, we couldn't accept it because _we don't have a meaningful, community accepted definition of what "done" looks like here._ I refuse to sign the maintainer team up for trying to hit a moving target.

I recommend that everyone that is interested in this [take the discussion to this Discuss topic](https://discuss.atom.io/t/what-will-it-take-to-implement-a-decent-regexp-engine-for-find-replace-in-atom/40956) and work out a definition of what a regular expression engine has to have to be acceptable. If a consensus can be reached, I will even create an Atom blog post to request further comments from the community to ensure that we are getting broad-based feedback and not leaving anyone out of the conversation. If that process results in consensus, then I would encourage someone to submit a PR that solves that problem and will work with the development team to get it accepted.

I want us to improve Atom and if we can be sure that the change won't just be making one group happy while antagonizing another, I'll stand behind it and advocate for it.

**Comment by lope** [comment]
@lee-dohm if you are concerned that a better RegExp engine like Oniguruma would be slow enough to upset users who don't need the features it provides, then I think the ideal upgrade to Atom's RegExp find/replace would be an option to select which RegExp engine to use.

@lee-dohm I think it's great that you've raised the performance issue of RegExp engines. And it's great that you don't want to make a change that will upset some users and make other users happy.

I think it's wrong that we should all discuss what RegExp features we want and collectively compromise on features and performance. That is pointless. You can't please everyone that way. But by letting people select their RegExp engine, everyone can be happy 😄 

So the fastest JS RegExp engine can be the default, and if they enter for example a lookbehind expression then it can let them know they're using the JS engine but they can switch to Oniguruma (or whatever) if they wish.

**Comment by lope** [comment]
@50Wliu I respect the Atom team's preference to work on startup time and performance.
But personally the startup time is plenty good enough for my liking. I don't reboot every 5 minutes so I don't care if atom opens in 1 second or 10 seconds. I've got a decent machine (dual core i7 2###, 4 threads)
I've scripted the startup of my whole dev environment. I make coffee while it all loads up.

Spending time to write convoluted RegExp expressions regularly throughout the day, because there is no lookbehind in JS RegExp is a much greater issue for me.

**Comment by MartinBonner** [comment]
@lope Noooooo!  Options are bad.  Collective compromise is a _much_ better solution.  I don't have time to learn about all the different engines and which one provides the features/performance I need.  (By all means have a fast engine which offers limited features and then _automatically_ switch to a slower engine if a more complex RE is used.)

(RegEx101.com is a very different case.)

**Comment by lope** [comment]
@MartinBonner collective compromise is the dumbest idea ever. If we go with your idea all cars would be one color. We wouldn't have sexual orientations that we choose but all have hermaphrodite partners. There would only be one item on the menu in the restaurant. You only have 1 channel on TV. You must just learn to like what was decided by the group. Say goodbye to your musical preferences, you only listen to one radio station now, selected by the group. And your clothing, your hairstyle. What type of mobile phone you use, what apps you have. What software is installed on your computer. How many GB of bandwidth you must pay for. Say goodbye to your choices there too. Don't impose your small-minded limited world on others.

I think automatic selection of the fastest RegExp engine is feasible, and by all means it can be the default option, but there must be an option, for people who do care.

You don't have the time to learn which RegExp engine supports the features you use, but you took the time to learn RegExp, and you take the time to learn to write software, and you take the time to comment on a thread about improving Atom's RegExp engine. What a joke.

**Comment by MartinBonner** [comment]
@lope:  There are obviously limits.  In my view Atom is too far towards the "knobs and buttons" end of the spectrum - I just want an editor that works, not one I have to program my self (if I wanted that, I know where to find Emacs).  Feel free to have the last word...

**Comment by lope** [comment]
![image](https://cloud.githubusercontent.com/assets/2905035/24289262/2bb21670-1089-11e7-9362-a33210059345.png)

**Comment by lope** [comment]
If you guys want a workaround to enable lookbehind right now in Atom, first close all atom windows, then run
`atom --js-flags="--harmony_regexp_lookbehind"`

I see searching the entire project with a lookbehind is spectacularly slow. It's a pity the engine doesn't somehow rewrite or optimize the search.
Slower than slow `(?<!O)\.foo[^A-Za-z]`
Near instant `[A-NP-Za-z]\.foo[^A-Za-z]`

So that's my feedback, for what it's worth. Harmony on 1.14.3 (will upgrade now)

**Comment by lee-dohm** [comment]
@lope your comment was deleted as a violation of the [Atom Code of Conduct](https://github.com/atom/atom/blob/master/CODE_OF_CONDUCT.md) as is insulting or derogatory. You may consider this an official warning.

**Comment by lee-dohm** [comment]
@lope This isn't up for debate. You have a couple options here:

1. You can work with others and compromise on what works best across all Atom users
1. You can fork the find-and-replace package and install whatever regular expression engine you want that works best for you

The Atom maintainer team is not going to be creating a pluggable architecture for regular expression engines. And we're going to continue to work on performance of various types including startup performance.

**Comment by laike9m** [comment]
NEED this feature!!! PLEASE!!!

**Comment by lee-dohm** [comment]
@laike9m Please read the [call to action above](https://github.com/atom/find-and-replace/issues/571#issuecomment-288859566) if you'd like to help define what "better" means.

**Comment by lope** [comment]
@lee-dohm From now on I won't express any disagreement or point out the flaws in people's logic because you interpret that as an insult or derogatory. So I'll just say have a nice day! Doesn't seem constructive, but you're the boss.

**Comment by jeancroy** [comment]
> There is no definition of performance expectations for both speed and memory (because I'm certain that will be the first complaint even if we select an engine that has all the features everyone wants)

> If you'd like to help define what "better" means.

As far as my issue is concerned (#557) better means "hang-proof" the editor, regardless of user input.
Regex are hard and it's easy to create exponential complexity mini program.

This criteria would be met by using JS engine in a worker thread, for example. Although that may require similar amount of work that the current Omnigumnura request.

**Comment by lee-dohm** [comment]
@jeancroy You make a good point. I don't think fixing that problem needs to wait for people to decide what regular expression features are most important, especially since it is something that I believe all regex implementations are going to face. So I'd like to keep that as a separate issue.

**Comment by MattSturgeon** [comment]
It's my interpretation that the general consensus seems to be that "better" or "fixed" could simply mean use Oniguruma for all regex stuff in Atom.

I believe the two main arguments for this have been 1) Atom already uses Oniguruma elsewhere, so it makes sense to be consistent and 2) Oniguruma adds requested features.

I would propose that if Oniguruma is advanced enough (and fast enough) for other Atom components, then it is likely to be good enough for Find, too.

Does anyone know of any reasons **not** to use Oniguruma for find and replace - or any reasons why it wouldn't be suitable as a universal regex engine for Atom?

**Comment by ErikCorryGoogle** [comment]
Lope, I tested your regexps in plain V8 and I get the lookbehind version to be 2-3 times slower. That doesn't fit very well with your "instant" vs. "slower than slow", but it's not satisfying.

I have submitted a patch to fix the performance: https://codereview.chromium.org/2777583003

**Comment by MartinBonner** [comment]
@MattSturgeon The obvious reason not to use Oniguruma for find and replace is that it involves work to stop using the Javascript regex and use Oniguruma instead.  In particular, if the Javascript regex is going to acquire the requested features _anyway_, it may be better to just wait for that to happen.

I'm not convinced that is a good enough reason, but I want to remind people it exists.

**Comment by lee-dohm** [comment]
@MattSturgeon Here's what I see happening if we just accept Oniguruma:

1. Atom maintainers do a bunch of work to migrate everything to Oniguruma
1. People involved in this conversation all cheer
1. Some other group comes along and says it is too slow or doesn't have X feature
1. The whole argument starts all over again

We're not going to play Whack-a-Mole. I want to see a list of features and performance targets. I want to get broad-based community consensus that the list of features and performance targets is acceptable. I want us to find a regex engine that meets those features and targets. Then when someone comes along saying _that_ isn't good enough I can point to all the work we did and the rationales everyone had for the choices made. "We couldn't think of a reason not to at the time," isn't a convincing rationale.

**Comment by Stanzilla** [comment]
What is wrong with ripgrep then? grep is a pretty accepted standard usually?

**Comment by jeancroy** [comment]
Mostly ripgrep is an alternative to https://github.com/atom/scandal and don't relate much to the regex engine used in buffer.

**Comment by bric3** [comment]
I have been monitoring this issue for a while and I'm sad of the state of the progress on this issue. I'm trying to make an honest feedback and I am trying to be constructive. I am myself a developer of an opensource framework, I know there is features wanted by some users that are not in the priority list for various reasons. While priorities are important let's not ignore features that are useful for the users. As a library developer several time I never imagined some usage that users had, and limitations in my library hit them, so I am enclined to listen about usage and help them if it is possible, sometime it can take a long time but we do it. Regarding the information this issue has, it is very limited it only have the _enhancement_ tag, whatever the Atom team chose it should at least indicate the state of this issue be it _delayed_, _stopped_, _re-scoped_, _low-priority_, _never_, _to be redefined_...

Feedback :

I am a strong Regex users, and I am usually disappointed by tools that don't support powerful enough edition. Especialy when there is complicated pattern to find (wiht big data it matter). Regarding regex I usually expects the features found in perlRE. Now I usually code with JetBrains tools that are very very good as IDEs. This lack of search support drives me and my colleague to use JetBrains IDEs and as such away from Atom. So we have a solution that works for us, but it is disrupting, and instead of starting with Atom we avoid it.

Some ideas :

* Why not [the silver searcher (`ag`)](https://github.com/ggreer/the_silver_searcher) which seems to follow the fairly common `ack` _standard_. Now I don't know if this approach can fit in the atom design as it requires to run a separate process. But the benefit of this approach allows the UI to not hang.

* Regex engines varies in features and performance, yet I think a fairly good scope (for a _better_ regex engine https://github.com/atom/find-and-replace/issues/571#issuecomment-288859566) would be the [perl regex](https://perldoc.perl.org/perlre.html) which has become quite popular outside the perl language (Java for example follows this engine in features).

**Comment by jeancroy** [comment]
> I am a strong Regex users, and I am usually disappointed by tools that don't support powerful enough edition. Especialy when there is complicated pattern to find (with big data it matter).

Can you share real life example where javascript engine fail your need ? I think at this point this would be the most useful input to tip the balance.

Imo Big data is a separate problem. With large enougth data it become more practical to write your own small tool than try to fit within the limitation of a general purpose text editor.

> Why not the silver searcher (ag). 

In general grep like tool lack the ability to do multiline matches. This is because speed-usability compromise is different. See for example https://github.com/BurntSushi/ripgrep/issues/176 or https://github.com/atom/scandal/issues/5.

Ag in particular is free of this issue, but seems to be less compatible with window (?).
Also atom has an instant preview feature that need to test regexp on only some part of file, or files that are not yet saved to disk. Finally the issue of real life limitation remains.

**Comment by ErikCorryGoogle** [comment]
My performance fix for lookbehinds (and lookaheads) in V8 has landed.  Of course it will take some time to percolate through to Atom.

**Comment by bric3** [comment]
@jeancroy Sorry for delayed answer.

> Can you share real life example where javascript engine fail your need ?
> I think at this point this would be the most useful input to tip the balance.

When dealing with a lot of heterogenous data, it is useful to inspect data. There are powerful tool on the comand line. But when you have to discover the data, and something is wrong or you want to just look at the structure of the data, I use quite often regular expressions, and more than rarely I use (positive / negative) look-ahead or look-behind expressions to identify specific parts of these files.
Actions may vary for me, e.g. sometime it is just looking up in a file. Sometime I really want to fix it.
What is disrupting is when you have to switch tools. Disclaimer : I understand what it is to use the right tool for the job, but for exploring data for any reason I'd like to stay in my editor.

----------

Ah yes your are correct it seems [complicated](https://github.com/ggreer/the_silver_searcher/wiki/Windows) to have ag on windows. There's yet another alternative to ag, it's written in go and called the platinium searcher (pt) :  https://github.com/monochromegane/the_platinum_searcher
As it is writen in go it is multipatform.

**Comment by jinglesthula** [comment]
I tried the `atom --js-flags="--harmony_regexp_lookbehind"` approach and now using lookbehind doesn't show the error, but it does sit and spin forever with "0 paths searched".  Searching w/o lookbehind in the regex still works.  Anyone else try this with better results?

More on topic, does anyone know why lookbehind didn't land in ES6?  That seems like the root problem.  Tried googling but didn't find much discussion.

**Comment by ErikCorryGoogle** [comment]
If it spins forever you may have written a regexp that takes forever. This is quite easy to do. Or you may have hit a bug.  Can you share the regexp in question?  There are people in this very thread who have tried the flag and it worked for them.

As for why it was not part of ES6, this is because there was not agreement on what form it should take. Perl-style fixed width lookbehinds had been proposed and I was not very interested in that happening. In the end I helped push the current full lookbehind proposal through so that we would not end up stuck with the fixed width version which has caused much gnashing of teeth in perl.  To get something on the TC39 standard you need a champion who wants to write the spec and argue for it in committee, and these days you also need to actually implement it.  Sometimes this takes time.

How do people feel about inline flags in regexps?  eg
/(foo(?i)bar)/
would match foobar and fooBAR but not Foobar because the /i flag only goes from the (?i) to the close parenthesis.   Is this something people miss?

**Comment by jeancroy** [comment]
> How do people feel about inline flags in regexps? eg
/(foo(?i)bar)/
would match foobar and fooBAR but not Foobar because the /i flag only goes from the (?i) to the close parenthesis. Is this something people miss?

I'm not sure I'd have much use for it. But if a single syntax is adopted I like the modifier on non-capturing group syntax. `/(foo(?i:bar))/` . It seems .Net handle it (actually both) and I like how the scope is well defined.

**Comment by whoinoi** [comment]
Is there any update on this? I found, that the following regular expression also doesn't work in Find in Projects: `(?<!\$)\$(?!\$)`

It says "Invalid regular expression `/(?`"

**Comment by fiapps** [comment]
In the interest of producing a concrete requirement, I (like many) require lookbehind, and in particular I want them to work for searches of the whole project. I would also point out that this was the requirement expressed in the original title of this issue: "positive and negative lookbehind (lookaround)." It only became a vague "better" due to a later edit.

Using Atom 1.25.0, with `atom --js-flags="--harmony_regexp_lookbehind"` to enable the experimental feature, I can search a single file quickly for `(?<!{)foobar\b`, but searching the entire project (23 files totaling less than 13000 lines) either takes an extremely long time or hangs, because after more than an hour, the spinner still said 0 files had been searched.

Over two years have passed since the [announcement](https://v8project.blogspot.it/2016/02/regexp-lookbehind-assertions.html) that lookbehind support was coming to V8, and we still don't have a way to reliably use lookbehind to search a whole project. Therefore, I do not think waiting for V8 to solve the problem is an adequate solution.

**Comment by naomik** [comment]
Sorry for the subjective title in the first place! Still waiting on lookaround in Atom...

**Comment by ErikCorryGoogle** [comment]
Is this really the regexp you used to test? If searching 13000 lines takes more than an hour then there's no way that's a V8 regexp performance issue.  You probably need to file a new bug.  Any way to work out which version of V8 this is?

I just tested this on V8 version 6.2 and the two regexps

/(?<!})foobar\b/

and

/foobar\b/

are exactly the same speed, about 14 million lines per second.

There is no reason for Atom to delay switching this on, that I can see.  The proposal is finished: https://github.com/tc39/proposals/blob/master/finished-proposals.md

**Comment by ErikCorryGoogle** [comment]
Atom has had lookahead since forever, so the new title is not very accurate.

**Comment by fiapps** [comment]
I actually used a very similar regex with another word in place of foobar. However, testing `(?<!{)foobar\b` I get the same results (now with Atom 1.25.1): I can search a single file almost instantly, but a search of the whole project hangs or takes an extremely long time. I don't know whether the bug is in V8 or Atom. I haven't done any Atom development, so I don't know how to find out where it hangs, but it looks like jinglesthula reported the same issue above.

As for the version of V8, I see that the [Atom 1.25.0 release notes](https://github.com/atom/atom/releases/tag/v1.25.0) report that Electron was upgraded to 1.7.11, and the [Electron 1.7.0 release notes](https://github.com/electron/electron/releases/tag/v1.7.0) report that V8 was upgraded to 5.8.283.38.

**Comment by ErikCorryGoogle** [comment]
V8 version 5.8 is now over a year old, from March 20 2017: https://v8project.blogspot.dk/2017/03/v8-release-58.html

Unfortunately the performance fix was done on March 28 2017, so it has not got through to Atom yet.

Nevertheless there must be a different issue here, since the improvement was only by a factor of 2-3, and you are seeing a factor of at least 1 million slowdown.

**Comment by mrmass** [comment]
In the meantime I still get invalid regular expression on lookbehind syntax... Were there any plans to fix this issue?

**Comment by billybooth** [comment]
Right, is there any traction on this in the release version?  I skimmed the thread, but I'd prefer to have negative lookbehind with poor performance than not at all.
